feat(elasticloadbalancingv2): add load balancer lookups

misterjoshua opened the pull request:
Adds `fromLookup()` methods to both Application and Network load balancers as well as their listeners.

Closes #11088 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 0f2a454979159a45e5f542c74a943d7e0fc2fd06
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=335753c4-48cb-465c-af32-7867a9c2737f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 1215ffb230e439e420bef4275115aa09014c25a7
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=459be8a1-f666-4758-ba19-6a414e996471%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 5efac1b4c7347da73736b096612132987d905ba5
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=cdbb7d33-a352-4a62-b549-5401489b86c7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c0175d67b9c6fd840881fd7160d88b6d7c4bbc73
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b2ca600f-3724-42e9-b19f-dccd1438fff0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: b372c874bc4ac4ac6489ef1521e8a79b5e5f67a0
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ed9e8903-899d-415f-b392-53e365bfb265%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d2ab397427472de09b43b46feb60c698a60e33f9
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b945a346-06cf-402f-8d98-5ee4465ef23c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 9c1ae33d6982e39e047774b412ca2ca9de893f09
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5b00aa9e-1a68-48f9-8383-5754e1c22cf4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: cf4da2d740d3643aee5c686dbd26518a1e6bfe8c
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=338ace19-b631-495c-bf6a-4c69a72cb398%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 5e5ffeff45a90fe9dbdd33c208deec672e4ec121
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4401b628-08d4-4862-8931-b9ccf5b4b914%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

misterjoshua commented:
I've looked over my code since I put this PR up. Given that this change entails a new manifest schema version, and that I have implemented looking up security groups by id: Is now a good time to add tag search for security groups?

njlynch commented on packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.ts:
```diff
@@ -86,12 +87,52 @@ export interface NetworkListenerProps extends BaseNetworkListenerProps {
+   * Looks up a network listener
+   */
+  public static fromLookup(scope: Construct, id: string, options: NetworkListenerLookupOptions): INetworkListener {
+    let listenerProtocol: cxschema.LoadBalancerListenerProtocol | undefined = undefined;
```
```suggestion
    let listenerProtocol: cxschema.LoadBalancerListenerProtocol | undefined;
```

njlynch commented on packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.ts:
```diff
@@ -86,12 +87,52 @@ export interface NetworkListenerProps extends BaseNetworkListenerProps {
+      loadBalancerType: cxschema.LoadBalancerType.NETWORK,
+    });
+
+    class LookedUp extends Resource implements INetworkListener {
+      public listenerArn = props.listenerArn;
+    }
```
This seems like a case where following a pattern and having commonality between the ALB and NLB interface yields a sub-optimal result. `NetworkListenerLookupOptions` extends `BaseListenerLookupOptions`, which includes `listenerArn` as one of the filtering options, but there's no case here for why someone would use `fromLookup` with `listenerArn` over `fromNetworkListenerArn`.

misterjoshua replied:
@njlynch Makes sense. Is your suggestion then to entirely separate the interfaces for `ApplicationListener.fromLookup()` and `NetworkListener.fromLookup()`? Or to push `BaseListenerLookupOptions.listenerArn` down to `ApplicationListenerLookupOptions.listenerArn`? Or...?

njlynch replied:
Well, as a bit of devil's advocate, what use case does adding the lookup by ARN enable? What specific functionality for `ApplicationListener.fromLookup()` is benefitted from `ApplicationListener.fromLookup({ listenerArn: 'foo' })` vs `ApplicationListener.fromApplicationListenerAttributes()` with the ARN?

misterjoshua replied:
@njlynch Fair enough comments. But I had in mind that if the user has the ARN of the listener, it's nice to have CDK get the security groups of the ALB by lookup as well. For instance, the user is probably going to need the ALB security groups to let the ALB traffic into an ECS/Fargate task.

They can certainly use `ApplicationListener.fromApplicationListenerAttributes()` with the ids of all involved resources, but then they'll need to provide the security groups of the load balancer as well. So, the improvement of lookup by ARN is primarily a convenience over needing to specify the security group.

If NLB's listener's interface had `ec2.IConnectable`, I'd probably say the same for it, but only `IApplicationListener` extends the interface. So ALB and NLB listeners are definitely a little incongruous in that respect.

njlynch replied:
Fair point. I can see the benefit of having the security groups automatically loaded. Perhaps we move `listenerArn` up to `ApplicationListenerLookupOptions` then. I'm also going to ask one of my colleagues to take a look as this PR since it's so big, and see if they have an opinion.

skinny85 replied:
I like moving `listenerArn` to `ApplicationListenerLookupOptions` if providing the ARN doesn't make sense for Network Listeners.

We can always move it back down, but we can't remove it after it's added to `NetworkListenerLookupOptions` 🙂.

njlynch commented on packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener.ts:
```diff
@@ -517,36 +545,46 @@ export interface ApplicationListenerAttributes {
   public readonly listenerArn: string;
 
-  constructor(scope: Construct, id: string, props: ApplicationListenerAttributes) {
+  constructor(scope: Construct, id: string, props: ImportedApplicationListenerBaseProps) {
     super(scope, id);
 
+    this.connections = props.connections;
     this.listenerArn = props.listenerArn;
```
I think a cleaner approach would be to keep `connections` and `listenerArn` abstract in this base class, and just set them in the concrete classes. Saves needing to create the `ImportedApplicationListenerBaseProps` and awkward `super` calls.

njlynch commented on packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener.ts:
```diff
@@ -106,12 +108,38 @@ export interface ApplicationListenerProps extends BaseApplicationListenerProps {
+    const props = BaseListener._queryContextProvider(scope, {
+      userOptions: options,
+      loadBalancerType: cxschema.LoadBalancerType.APPLICATION,
+      listenerProtocol: options.listenerProtocol == ApplicationProtocol.HTTP
```
```suggestion
      listenerProtocol: options.listenerProtocol === ApplicationProtocol.HTTP
```

Please use `===` and `!==` throughout.

njlynch reviewed (changes requested):
Thanks for the contribution! I would have preferred slightly smaller chunks over something this size -- e.g., just the load balancers by ARNs/tags would have been significant in itself. That being said, we'll take this as-is. I did a first pass of comments, will go deeper on the next round.

misterjoshua commented:
Thanks for the review. I'll get to these changes as soon as possible. :)

misterjoshua commented on packages/@aws-cdk/aws-ec2/lib/security-group.ts:
```diff
@@ -294,6 +296,31 @@ export interface SecurityGroupImportOptions {
+      allowAllOutbound: true,
+    };
+
+    const attributesToUse = attributes ?? dummyAttributes;
```
Only because I found something similar in `Vpc.fromLookup()`, where the dummy value is held separately and not given as `dummyValue` input to `getValue`. Given that I don't have strong feelings here, I'll change this to just use `dummyValue` flat out.

njlynch replied:
The VPC is the odd one out. There was a specific subnet-related problem there that needed to be special-cased (see #4090) .

misterjoshua commented on packages/aws-cdk/lib/context-providers/load-balancers.ts:
```diff
@@ -0,0 +1,292 @@
+    for (const loadBalancer of pageItems) {
+      loadBalancers.push(loadBalancer);
+    }
+  } while (page.NextMarker);
```
Whoops. Good catch!

misterjoshua commented on packages/aws-cdk/lib/context-providers/load-balancers.ts:
```diff
@@ -0,0 +1,292 @@
+function tagsMatch(tagDescription: AWS.ELBv2.TagDescription, requiredTags: Record<string, string>) {
+  const tagsByName: Record<string, string> = {};
+  for (const tag of tagDescription.Tags ?? []) {
+    tagsByName[tag.Key!] = tag.Value!;
```
@njlynch Hmm. I see. If I change the code as follows, I can handle an undefined `Value` and the loop conditional will work with strict not equals.

```ts
const tagsByName: Record<string, string | undefined> = {};
for (const tag of tagDescription.Tags ?? []) {
  tagsByName[tag.Key!] = tag.Value;
}
```

However, unless we change `cxschema.Tag`, we can't match on an undefined tag value. From a quick glance, this change would entail a few changes to `CloudFormationStackArtifact`. What do you think we should do?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 5dfc521c93a03803f1b79ee196a47a2451016c3f
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a8fc23d8-58e7-40c1-80f0-dc1adeedaf6a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

misterjoshua commented on packages/@aws-cdk/aws-ec2/lib/security-group.ts:
```diff
@@ -294,6 +296,31 @@ export interface SecurityGroupImportOptions {
+  /**
+   * Look up a security group by id.
+   */
+  static fromLookup(scope: Construct, id: string, securityGroupId: string) {
```
This is in.

misterjoshua commented on packages/aws-cdk/lib/context-providers/load-balancers.ts:
```diff
@@ -0,0 +1,292 @@
+/**
+ * Find load balancers by the given filter args.
+ */
+async function findLoadBalancers(elbv2: AWS.ELBv2, args: cxschema.LoadBalancerFilter) {
```
I've implemented these checks higher up because the context providers know better.

misterjoshua commented:
@njlynch Hey. I think I've got most of your review items addressed in the latest batch of commits. There are two outstanding things as far as I can tell:

* The use case for listener lookups by ARN. (My response is in the review thread.)
* Need to figure out what to do about `cxschema.Tag.value` not allowing undefined. (Comments in the review thread.)

In the meantime, I'll keep an eye out for any more feedback.

misterjoshua commented on packages/aws-cdk/lib/context-providers/load-balancers.ts:
```diff
@@ -0,0 +1,292 @@
+  /**
+   * Finds the first matching listener from the list of load balancers.
+   */
+  private async findFirstMatchingListener(elbv2: AWS.ELBv2, loadBalancers: AWS.ELBv2.LoadBalancers, query: LoadBalancerListenerQuery) {
```
This behavior is now updated.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c4e9973c8e534677cf5dcbd3e857b686a1754103
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=eeff4f8a-5396-4207-8e47-c733e966654a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 9a34689b0f7313fda45f6d7d0bd1580514309a09
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=359b44c9-2039-4055-8e95-055a677001ca%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 34f870372ce6b3e403e3602f8f40fbeee6302c27
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=cfcf2132-068a-4365-86a9-ac8085d23b2f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

misterjoshua commented:
After the merge, the build broke. It looks like I may need to add another allowed breaking change. JSII 1.14 notices the three new query types in a new way:

```
[weakened:@aws-cdk/cloud-assembly-schema.MissingContext]
```

From the output, jsii seems to indicate that the new `MissingContext` can't be assigned to the old `MissingContext`. This is to be expected, considering that we've added three arms to the context queries union type and have updated the schema major version.

I see that 1.71.0 is in the wings, so I'll wait to merge that in before I push up another build.

misterjoshua commented on allowed-breaking-changes.txt:
```diff
@@ -47,4 +47,12 @@ incompatible-argument:@aws-cdk/aws-ecs.Ec2TaskDefinition.addVolume
+# | LoadBalancerListenerContextQuery
+# | SecurityGroupContextQuery
+# These additions changed the signature of loadAssemblyManifest()
+change-return-type:@aws-cdk/cloud-assembly-schema.Manifest.loadAssemblyManifest
```
Changing line 58 to the following fixes the problem with this PR's build, introduced when merging in JSII 1.14 from master:

```
weakened:@aws-cdk/cloud-assembly-schema.MissingContext
````

skinny85 commented on packages/@aws-cdk/cx-api/lib/context/security-group.ts:
```diff
@@ -0,0 +1,16 @@
+  readonly securityGroupId: string;
+
+  /**
+   * Whether the security group of the load balancer allows all outbound
```
`SecurityGroup.fromLookup()` can also be called without a LoadBalancer context... what does this property mean then?

misterjoshua replied:
It's more general-purpose than this comment conveys. I'll adjust the wording so that this is clear.

skinny85 reviewed (commented):
Looks fine to me. Couple of minor issues that I've found.

Since this is such a complicated change, I would loop in @rix0rrr to sanity-check only the context provider changes look good, and @iliapolo just to verify the cx-api bump. Thanks guys!

misterjoshua commented on packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener.ts:
```diff
@@ -106,12 +108,38 @@ export interface ApplicationListenerProps extends BaseApplicationListenerProps {
+    const props = BaseListener._queryContextProvider(scope, {
+      userOptions: options,
+      loadBalancerType: cxschema.LoadBalancerType.APPLICATION,
+      listenerProtocol: options.listenerProtocol === ApplicationProtocol.HTTP
```
Bug. My bad. I'll fix. I intended that it should do as the docs say.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 8928287216a18b3b1d97821912ba572f04acd536
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9d81814d-2253-4866-ac90-411144f0ba15%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 6a4e369e527b220596e29ae55421969a900b246f
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=86b750af-e6ca-4573-814b-97083a8774a7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 8c2c7341b0e35c34ad79f05bb6772c1bff8d41f4
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=12d0a0e0-b1bd-4c19-ae01-07538a688bb7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

misterjoshua commented:
@skinny85 I believe I've addressed your review items in the latest commits. I also updated from the main branch. Everything seems to be in working order. I'll continue to keep an eye out for more changes. :)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d23f909502ffc1d5edcf525f8a05b7bc931f6bb8
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=64b78da9-0f53-4976-b532-8b8fb78ae118%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

iliapolo commented:
@skinny85 @misterjoshua I'm taking a look at the cx-api changes, added `do-not-merge` in the meantime just in case.

skinny85 commented:
Thanks @iliapolo !

@rix0rrr if you could take a quick look at just the provider changes, that would be great.

misterjoshua commented:
While you guys are reviewing, would you prefer that I resolve the merge conflicts or handle them for later?

njlynch commented:
> While you guys are reviewing, would you prefer that I resolve the merge conflicts now or handle them later?

Feel free to resolve them now.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: f4c84e5260cb20c4076527c0b2b3418fd2326de9
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fc9539c7-e55b-4337-b527-4294f8674524%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

iliapolo commented:
@njlynch @skinny85 I'm good with the cx-api changes, so removing the `do-not-merge` I added.

njlynch commented on packages/aws-cdk/lib/context-providers/load-balancers.ts:
```diff
@@ -0,0 +1,309 @@
+      throw new Error('The load balancer listener query must specify at least one of: `listenerArn`, `loadBalancerArn` or `loadBalancerTags`');
+    }
+
+    if (query.listenerArn) {
+      // When we know a listener arn, we can query for that listener directly.
+      return this.getListenerByArn(elbv2, query);
+    } else {
+      // When we don't know a listener arn, we need to find it by traversing
+      // all associated load balancers.
+      return this.getListenerByFilteringLoadBalancers(elbv2, query);
+    }
```
`~` (Style) I'd probably just inline this:

```suggestion
    return query.listenerArn ? this.getListenerByArn(elbv2, query) : this.getListenerByFilteringLoadBalancers(elbv2, query);
```

njlynch reviewed (commented):
Really close, just one legitimate comment and a few stylistics nitpicks.

misterjoshua commented on packages/@aws-cdk/aws-elasticloadbalancingv2/lib/shared/base-listener.ts:
```diff
@@ -1,12 +1,101 @@
+        loadBalancerArn: options.userOptions.loadBalancerArn,
+        loadBalancerTags: cxschemaTags,
+        loadBalancerType: options.loadBalancerType,
+      } as cxschema.LoadBalancerListenerContextQuery,
```
The props are a `{ [key: string]: any }`, so the cast isn't strictly necessary. But, I think the type hint is valuable if only for the reader's understanding, as otherwise, the correct type for the context query isn't obvious on inspection. I'll defer to your judgement on this matter though. :)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 2f2a992f52607d7939039b2ab6ad785ff937db6c
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=afdb45de-a430-43b9-97c1-430f93e4cc95%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 19dc7646466f3685b0365df89a59df273b827870
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2cb87d3a-80c2-45e4-a736-518e2d09347d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: e0d3b9cf1651c4a216bd9dabd769ddc816adeffd
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=352a21fd-2657-4f91-afa4-7ae277f87492%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

njlynch reviewed (approved):
Great! Thanks for the contribution!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 488b6d89b417d5f83b87faa325660b088bf7aeef
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=14f91036-dff8-419f-ab44-c4e4194fb73b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
